### ISSUE TITLE
IPNS+IPFS as a NoSQL database

### ISSUE THREAD

**Issue by rethore** [ipfs/notes#70]
It would be quite nice if we could have a library that would emulate a NoSQL database. So in essence offering an abstraction of IPNS+IPFS through methods such as `insert`, `update`, `remove` and `find`. From there we could easily build distributed web apps based on this as a back-end. So for instance meteor is using a combination of a client side minimongo library that syncs with a server side MongoDB. If we were to swap the MongoDB with this new library, and move most of the logic on the client side, we could achieve quite a bit already. Going further would require to run the server side as well locally, but it's not impossible, just less convenient. There is of course some problems arising of having an open access back-end database, but as long as people are not able to erase other peoples entries, their activity, if judge negative by the community, could be filtered out in a p2p fashion.

**Comment by jbenet** [comment]
@rethore yep, this is very possible. there are many systems that implement NoSQL dbs and even full SQL dbs on top of KV-stores. we should look at them and try them out.

one good route would be to start with things from the level nodejs ecosystem. there's a few SQL adaptors wrapping level.

some interesting links:
- https://www.npmjs.com/package/sql-engine
- https://github.com/LMDB/sqlightning
- https://sqlite.org/src4/doc/trunk/www/design.wiki
- https://sqlite.org/src4/doc/trunk/www/bt.wiki
- http://docs.oracle.com/cd/E17076_02/html/bdb-sql/dbsqlbasics.html#identicalusage

**Comment by whyrusleeping** [comment]
we should take a look at using the files api to build such a thing.

**Comment by jbshirk** [comment]
https://github.com/bigchaindb/bigchaindb
https://www.bigchaindb.com/features/
https://www.bigchaindb.com/whitepaper/  *ipfs mentioned*

**Comment by fazo96** [comment]
The problem is handling permissions. As far as writing and deleting it can all be implemented, but we need a system where each folder has different permissions. It will be possible when we can have more people allowed to publish a given IPNS record :+1:

**Comment by Kubuxu** [comment]
Can a directory with subdirectories being `/ipns/Qm....` be created?

**Comment by whyrusleeping** [comment]
@fazo96 btw, you can generate a key with [this](https://github.com/whyrusleeping/ipfs-key) and use it to publish an ipns entry anytime without the daemon with [this](https://github.com/whyrusleeping/ipns-pub)

This behaviour will be integrated into ipfs itself soon, as well as a mechanism for storing and managing keys. But those two programs should serve as a decent way to accomplish a few things now

**Comment by fazo96** [comment]
Wow that's very nice @whyrusleeping! I hope to see that in the HTTP API soon

**Comment by rethore** [comment]
thanks @jbshirk, I wasn't aware of this project. It does seem to complement IPFS and ethereum quite well. It would be nice to build a distribution stack that bundle those three projects together and integrate them nicely. Ideally the incentivization scheme should look at compensating for bandwidth, CPU and storage. A complete decentralised crypto-economy and internet stack. This is a really exciting prospects.

**Comment by tokyo-jesus** [comment]
I'm in the early stages of developing a graph database on IPFS. The idea is to publish metadata (graphs, vertices, edges, schemas) via IPNS as RWORSets of IPFS hashes. There are some performance considerations, but I'm thinking that GPG could do well as the backend to a cryptographic RBAC system.

I'm in a bit of a time crunch getting funding and such, but I'm excited to finally get cracking on [Infinite Automata] and collaborating with y'all starting this winter.

[Infinite Automata]: https://www.infinite-automata.com/

**Comment by fdietze** [comment]
@tokyo-jesus I'd like to know more details of your project. Like how would a db schema look like? Can there be something like transactions? How to avoid race conditions?

**Comment by tokyo-jesus** [comment]
Schemas consist of primitive types and named references to other schemas. CRDTs don't have a concept of locking, the state is guaranteed to be eventually-consistent, which is enough for most needs, but not all. Long term, we'd like to enable strongly-consistent transactions through another distributed mechanism that can provide locks.

**Comment by jbenet** [comment]
@tokyo-jesus would be great! 👍

**Comment by jbshirk** [comment]
relevant: https://www.confluent.io/blog/turning-the-database-inside-out-with-apache-samza/

**Comment by bitcard** [comment]
mark
